docs/proposals/kubermatic-installer.md

```diff
@@ -0,0 +1,128 @@
+
+1. The admin prepares a `KubermaticConfiguration` and a Helm `values.yaml`. The goal here
+   is to require as minimal customization as needed and to smartly make the right choices:
+   For example, things that need to be configured identically in both files can be
+   defaulted from one into the other.
+1. The admin uses the installer like so:
+
+    ```
+    export KUBECONFIG=...
+    # or use -kubecofig=...
+    ./installer deploy --config kubermaticconfig.yaml --values values.yaml
+    ```
+
+   And off they go! The installer will validate the configuration and perform all steps
+   to install the Kubermatic stack.
+1. Done.
+
+### Goals
+
+* Provide a command-line application for the installer and bundle it together with the
+  relevant Helm charts into a Docker image. Maybe even embed the charts into the installer
+  binary.
+* Use Helm 3 for managing releases.
+* Do not require the admin to manually use Helm to manage their installation.
+* Allow the admin to manually use Helm if they so choose. Do not hide things from an admin,
+  but make it clear that they are leaving the supported paths.
+* Provide clear, helpful log output and error messages.
+* Never perform dangerous (as in: can possibly lose data or change LoadBalancer IPs)
+  migration steps without warning the user first (e.g. "run again with --yes-migrate-x
+  to perform the migration").
+* Pin the installer to a specific Kubermatic version, i.e. to choose a different version
+  a different installer release must be downloaded.
+* Use the installer during E2E tests.
+
+### Non-Goals
+
+* Downgrades.
+* Interactive CLI/ncurses interface.
+* Migrate manual installations to be installer-managed.
+* Migrate Helm-based installations to the Kubermatic Operator.
+
+### Implementation
+
+The installation generally done sequentially. First experiments with doing it in parallel
+tasks yielded no good results:
+
+* It was difficult to print an understandable log for the user. Keeping the user informed
+  about what is happening to their cluster is important.
+* Error handling was difficult, especially in longer running parallel Helm invocations
+  and dependent things.
+* It makes reasoning when writing migration steps much harder if you have to very
+  carefully plan your migration code in all the goroutines that run during an upgrade.
+
+Having sequential steps made the code much easier to read and maintain. It sacrifices some
+speed for ease of use.
+
+During `./installer deploy`, the chosen "stack" is installed. By default (and currently
+because it's the only one) this is the "kubermatic" stack. The stack then does whatever
+it needs to run. For Kubermatic this means
+
+* setting up a "kubermatic-fast" StorageClass; this is a leftover from very early installer
+  versions and requires us to know the cloud provider. This requires yet another setting
+  to be made from the user, so maybe we should instead just inform the user that they need
+  to create a StorageClass?
```

Comment:
From the code there is already an implementation for the storageClass. I would say we keep it this way and determine which cloud provider the user is using. Obviously a user might install it somewhere else than one of the 3 major cloud providers (Azure, AWS, Gcloud) and in this case, we need to inform the inform the user that he needs to provision it manually.